File: codetandem/code_parser.py
"""
Code parser for extracting user-written code from source files.

This module provides functionality to locate TODO comments and extract
the code written by users to fulfill those tasks.
"""

import re
from pathlib import Path
from typing import Optional, Tuple


class CodeExtraction:
    """Represents extracted code from a source file."""

    def __init__(
        self,
        file_path: str,
        todo_line: int,
        todo_text: str,
        code: str,
        start_line: int,
        end_line: int,
    ):
        self.file_path = file_path
        self.todo_line = todo_line
        self.todo_text = todo_text
        self.code = code
        self.start_line = start_line
        self.end_line = end_line

    def to_dict(self):
        """Convert to dictionary representation."""
        return {
            "file_path": self.file_path,
            "todo_line": self.todo_line,
            "todo_text": self.todo_text,
            "code": self.code,
            "start_line": self.start_line,
            "end_line": self.end_line,
        }


def extract_todo_code(
    file_path: str, todo_text: Optional[str] = None, todo_index: int = -1
) -> Optional[CodeExtraction]:
    """
    Extract code associated with a TODO comment from a source file.

    Args:
        file_path: Path to the source file
        todo_text: Optional specific TODO text to search for
        todo_index: Index of TODO to extract (-1 for last, 0 for first, etc.)

    Returns:
        CodeExtraction object if TODO found, None otherwise
    """
    path = Path(file_path)
    if not path.exists():
        return None

    try:
        content = path.read_text(encoding="utf-8")
    except Exception:
        return None

    lines = content.splitlines()

    # Find all TODO comments
    todo_pattern = re.compile(r"//\s*TODO:?\s*(.+)", re.IGNORECASE)
    todos = []

    for line_num, line in enumerate(lines, 1):
        match = todo_pattern.search(line)
        if match:
            todos.append((line_num, match.group(1).strip()))

    if not todos:
        return None

    # Select the target TODO
    if todo_text:
        # Find TODO by matching text
        target_todo = None
        for line_num, text in todos:
            if todo_text.lower() in text.lower():
                target_todo = (line_num, text)
                break
        if not target_todo:
            return None
    else:
        # Use index
        try:
            target_todo = todos[todo_index]
        except IndexError:
            return None

    todo_line, todo_content = target_todo

    # Extract code after TODO until next TODO or end of meaningful code
    start_line = todo_line
    end_line = todo_line

    # Find the next TODO or end of file
    next_todo_line = None
    for line_num, _ in todos:
        if line_num > todo_line:
            next_todo_line = line_num
            break

    # Determine end of code block
    if next_todo_line:
        end_line = next_todo_line - 1
    else:
        end_line = len(lines)

    # Extract code lines (skip the TODO line itself)
    code_lines = []
    for i in range(todo_line, end_line):
        if i < len(lines):
            line = lines[i]
            # Skip the TODO comment line
            if i == todo_line - 1:
                continue
            code_lines.append(line)

    # Trim trailing empty lines
    while code_lines and not code_lines[-1].strip():
        code_lines.pop()
        end_line -= 1

    code = "\n".join(code_lines)

    return CodeExtraction(
        file_path=file_path,
        todo_line=todo_line,
        todo_text=todo_content,
        code=code,
        start_line=start_line + 1,  # Line after TODO
        end_line=end_line,
    )


def find_all_todos(file_path: str) -> list[Tuple[int, str]]:
    """
    Find all TODO comments in a file.

    Args:
        file_path: Path to the source file

    Returns:
        List of tuples (line_number, todo_text)
    """
    path = Path(file_path)
    if not path.exists():
        return []

    try:
        content = path.read_text(encoding="utf-8")
    except Exception:
        return []

    lines = content.splitlines()
    todo_pattern = re.compile(r"//\s*TODO:?\s*(.+)", re.IGNORECASE)
    todos = []

    for line_num, line in enumerate(lines, 1):
        match = todo_pattern.search(line)
        if match:
            todos.append((line_num, match.group(1).strip()))

    return todos

File: src/utils/code-parser.ts
/**
 * Code parser for extracting user-written code from source files
 */

import { readFile } from 'fs/promises';

export interface CodeExtraction {
  filePath: string;
  todoLine: number;
  todoText: string;
  todoId?: string; // Extracted from TODO: [id] format
  successCriteria?: string[]; // Extracted from SUCCESS CRITERIA comments
  code: string;
  startLine: number;
  endLine: number;
}

export async function extractTodoCode(
  filePath: string,
  todoText?: string,
  todoIndex: number = -1
): Promise<CodeExtraction | null> {
  try {
    const content = await readFile(filePath, 'utf-8');
    const lines = content.split('\n');

    // Find all TODO comments with optional ID
    // Supports: // TODO: [obj-1] Description
    //          // TODO: Description
    const todoPattern = /\/\/\s*TODO:?\s*(?:\[([^\]]+)\])?\s*(.+)/i;
    const todos: [number, string, string | undefined][] = [];

    lines.forEach((line, index) => {
      const match = line.match(todoPattern);
      if (match) {
        const todoId = match[1]?.trim();
        const todoText = match[2]?.trim() || '';
        todos.push([index + 1, todoText, todoId]);
      }
    });

    if (todos.length === 0) return null;

    // Select target TODO
    let targetTodo: [number, string, string | undefined] | undefined;
    if (todoText) {
      targetTodo = todos.find(([, text]) => text.toLowerCase().includes(todoText.toLowerCase()));
    } else {
      targetTodo = todos[todoIndex === -1 ? todos.length - 1 : todoIndex];
    }

    if (!targetTodo) return null;

    const [todoLine, todoContent, todoId] = targetTodo;

    // Extract success criteria from comments above TODO
    const successCriteria: string[] = [];
    if (todoId) {
      // Look for SUCCESS CRITERIA comments above the TODO
      let criteriaStartLine = -1;
      for (let i = todoLine - 2; i >= 0; i--) {
        const line = lines[i]?.trim() || '';
        if (line.startsWith('//') && line.includes('SUCCESS CRITERIA') && line.includes(todoId)) {
          criteriaStartLine = i;
          break;
        }
        // Stop if we hit non-comment or empty line
        if (!line.startsWith('//') && line !== '') {
          break;
        }
      }

      // Extract criteria lines
      if (criteriaStartLine >= 0) {
        for (let i = criteriaStartLine + 1; i < todoLine; i++) {
          const line = lines[i]?.trim() || '';
          if (line.startsWith('//')) {
            // Remove leading // and whitespace, and optional - prefix
            const criterion = line.replace(/^\/\/\s*-?\s*/, '').trim();
            if (criterion && !criterion.includes('TODO:')) {
              successCriteria.push(criterion);
            }
          }
        }
      }
    }

    // Find code after TODO
    const nextTodoLine = todos.find(([line]) => line > todoLine)?.[0];
    const endLine = nextTodoLine ? nextTodoLine - 1 : lines.length;

    // Extract code (skip TODO line)
    const codeLines = lines.slice(todoLine, endLine);

    // Trim trailing empty lines
    while (codeLines.length && !codeLines[codeLines.length - 1]?.trim()) {
      codeLines.pop();
    }

    return {
      filePath,
      todoLine,
      todoText: todoContent,
      todoId,
      successCriteria: successCriteria.length > 0 ? successCriteria : undefined,
      code: codeLines.join('\n'),
      startLine: todoLine + 1,
      endLine: todoLine + codeLines.length,
    };
  } catch {
    return null;
  }
}

export interface TodoInfo {
  line: number;
  text: string;
  id?: string;
}

export async function findAllTodos(filePath: string): Promise<TodoInfo[]> {
  try {
    const content = await readFile(filePath, 'utf-8');
    const lines = content.split('\n');
    // Supports: // TODO: [obj-1] Description or // TODO: Description
    const todoPattern = /\/\/\s*TODO:?\s*(?:\[([^\]]+)\])?\s*(.+)/i;
    const todos: TodoInfo[] = [];

    lines.forEach((line, index) => {
      const match = line.match(todoPattern);
      if (match) {
        const todoId = match[1]?.trim();
        const todoText = match[2]?.trim() || '';
        todos.push({
          line: index + 1,
          text: todoText,
          id: todoId,
        });
      }
    });

    return todos;
  } catch {
    return [];
  }
}
